Resteasy Reactive: Support use of optional with list/set/sortedset

Sgitario opened the pull request:
Support use when using query params with some collections. Example:

```java
@Path("/list")
    @GET
    public String sayHelloToList(@QueryParam("name") final Optional<List<String>> names) {
        return doSayHelloToCollection(names);
    }
```

fix https://github.com/quarkusio/quarkus/issues/23898

Sgitario commented:
/cc @geoand @Plawn

Sgitario commented:
Moving to Draft as this is not fully working yet.

geoand reviewed (approved):
Thanks!

Sgitario commented:
> Moving to Draft as this is not fully working yet.

Now, the PR is ready to be reviewed.

This is a controversial difference of behaviors here between Resteasy Reactive and Resteasy Classic. Let's imagine we have the following endpoint:

```java
@Path("/list")
@GET
public String sayHelloToList(@QueryParam("name") final Optional<List<String>> names) {
   // ...
}
```

If no query params are being set, I would expect to receive an optional empty. This is working like this in Resteasy Reactive, but in Resteasy Classic, you'll receive the optional with an empty list of strings.

geoand commented:
`Optional<List<T>>` is always weird if you ask me...

I think the way we have it now is fine, I wouldn't worry too much about maintaining compatibility on this particular point

Sgitario committed:
Resteasy Reactive: Support use of optional with list/set/sortedset

Support use when using query params with some collections. Example:

```java
@Path("/list")
    @GET
    public String sayHelloToList(@QueryParam("name") final Optional<List<String>> names) {
        return doSayHelloToCollection(names);
    }
```

fix https://github.com/quarkusio/quarkus/issues/23898

Plawn commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/OptionalConverter.java:
```diff
@@ -18,7 +19,14 @@ public Object convert(Object parameter) {
+            Object converted = delegate.convert(parameter);
+            if (converted != null
+                    && converted instanceof Collection
+                    && ((Collection) converted).isEmpty()) {
```
Is it really necessary to check if the collection is Empty ?

Sgitario replied:
To return optional empty when the collection is empty, yes. See my comment here: https://github.com/quarkusio/quarkus/pull/24055#issuecomment-1056810692

Plawn replied:
Oh I see, you are right, It's better this way

michalszynkiewicz commented:
Should we support `Optional<List>`, etc in the client too? Hmmm... maybe we already do

Sgitario commented:
> Should we support `Optional<List>`, etc in the client too? Hmmm... maybe we already do

As far I could see, the rest client reactive extension already supports it.
